rfcs/020 - Dependencies permissions RFC.md

```diff
@@ -0,0 +1,193 @@
+# 020 - Dependencies permissions RFC
+
+## Current Status
+
+### Proposed
+
+2021-08-05
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- David Ellis <david@alantechnologies.com>
+- Luis De Pombo <luis@alantechnologies.com>
+- Colton Donnelly <colton@alantechnologies.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Alejandro Guillen <alejandro@alantechnologies.com>
+
+## Summary
+
+Alan's third-party module permission system is one of the value propositions of the language. It adds a layer of security no other project approaches. The idea is to allow users to prevent specific third-party dependencies from having access to specific standard libraries that they should not have access to. This can be achieve with current mocking built-in and some updates to the `@std/deps` standard library.
+
+The priorities of the module resoulution system would go as follow starting with the highest priority:
+
+1. Actual dependency level `modules` directory
+2. Global dependencies `modules` directory
+3. Application `modules` directory
+
+Meaning that inner `modules` > outer `modules`.
+
+## Expected SemBDD Impact
+
+This would be a minor update if we were post-1.0 as it should have zero breaking impact on existing code.
+
+## Proposal
+
+This proposed API takes current module resolution and built-in mock system as an advange. The main goal is to have applications with a **safe-by-default** behavior on install. The result will be an API where users explicitly choose which standard libraries they will be using. Additionally, external libraries can be added and specific blocks will be applied if necessary for each one.
+
+The API will handle the following actions:
+
+- Define which **standard libraries** will be used in the Alan application/library. Will translate in a removal from the "blacklist".
+
+- Define which dependencies or standard libraries will be disabled for all dependencies. Meaning if I'm installing 2 external dependencies, and both of them use `@std/cmd` and I want to block it, I can `disable` `@std/cmd` and will be block for both of them. It will be translated in mocks at `./depeendencies/modules/` level.
+
+![Block at library level](./lib-lvl-v1.png)
+
+![Block at app level](./app-lvl-v1.png)
+
+- Define which external libraries will be part of the project and we could add specific blocks for dependecies used in the external library.
+
+The `@std/deps` will have the following API:
+
+```ts
+type Package {
+  std: Array<string>,
+  dependencies: Array<Dependency>,
+  block: Array<string>,
+  fullBlock: Array<string>,
+}
+
+type Dependency {
+  url: string, // Empty string or ignore it for global blocks at application level
+  block: Array<string>,
+  fullBlock: Array<string>,  // fullBlock, deepBlock, extensiveBlock, exhaustiveBlock?
+  package: Package,
```

Comment:
Why does this property exist?

Reply:
In order to distinguish `block` from `Dependency` and `Package`. When returning `Dependency` we need a way to keep track of the `Package` object.

Reply:
Got it.